KAFKA-5226: Fixes issue where adding topics matching a regex

bbejeck opened the pull request:
subscribed stream may not be detected by all followers until
onJoinComplete returns.

bbejeck committed:
KAFKA-5226: Fixes issue where adding topics matching a regex
subscribed stream may not be detected by all followers until
onJoinComplete returns.

bbejeck commented:
This PR address situation where adding topics matching a regex subscribed stream may not be detected by all followers in `StreamPartitionAssignor.subscribe` but are picked up when 
`ConsumerCoordinator.onJoinComplete` returns.

ping @mjsax @dguy @guozhangwang  for review

bbejeck commented:
The test failure is `kafka.admin.DescribeConsumerGroupTest.testDescribeExistingGroupWithNoMembersWithNewConsumer` so I think this is unrelated to streams and this PR

bbejeck commented:
retest this please

guozhangwang commented:
Good find!

About the fix itself, in `ConsumerCoordinator` we already have the logic for handling unknown topics coming from the leader's assignment:

```
// check if the assignment contains some topics that were not in the original
        // subscription, if yes we will obey what leader has decided and add these topics
        // into the subscriptions as long as they still match the subscribed pattern
        //
        // TODO this part of the logic should be removed once we allow regex on leader assign
        Set<String> addedTopics = new HashSet<>();
        for (TopicPartition tp : subscriptions.assignedPartitions()) {
            if (!joinedSubscription.contains(tp.topic()))
                addedTopics.add(tp.topic());
        }

        if (!addedTopics.isEmpty()) {
            Set<String> newSubscription = new HashSet<>(subscriptions.subscription());
            Set<String> newJoinedSubscription = new HashSet<>(joinedSubscription);
            newSubscription.addAll(addedTopics);
            newJoinedSubscription.addAll(addedTopics);

            this.subscriptions.subscribeFromPattern(newSubscription);
            this.joinedSubscription = newJoinedSubscription;
        }

        // update the metadata and enforce a refresh to make sure the fetcher can start
        // fetching data in the next iteration
        this.metadata.setTopics(subscriptions.groupSubscription());
        client.ensureFreshMetadata();
```

I think a better way to fix it is to modify `StreamPartitionAssignor#onAssignment` which is called after the above code and hence when it is called the metadata should have been refreshed. And in that call we can trigger `streamThread.builder.updateSubscriptions()`.

Also if we do that in `onAssignment`, maybe we do not need to call `updateSubscription` in `StreamPartitionAssignor#subscription()` any more?

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/integration/RegexSourceIntegrationTest.java:
```diff
@@ -322,6 +323,61 @@ public void testShouldReadFromRegexAndNamedTopics() throws Exception {
+
+        try {
+            streamsLeader.start();
+            Thread.sleep(1000);
```
I'd suggest avoid `Thread.sleep()` as much as possible in unit test as it will introduce timing-dependent flakiness. If there is any condition that we need to wait on consider using `TestUtils.waitForCondition`.

Also this test seem not always testing the race condition that can cause the issue; I'd suggest make it really "modular" by adding a test for `StreamThread`, in which we do not need to start the thread but just mocking its embedded `StreamPartitionAssignor#onAssignment()` which gives it some assigned partition whose topic is not unknown, and then call its `onPartitionAssigned` callback to check if the topology builder has updated its topology.

bbejeck replied:
I'll remove the first two, but ok to leave in the one on line 352?  I want to space the the creation of topics.

guozhangwang replied:
I'm actually thinking more about making the unit test really "unit" by mocking stream partition assignor to always return unknown topic partitions; the current test case seems to be more "integration test"?

bbejeck replied:
Yes it is, I see your point now about the test, I'll make the changes.

bbejeck commented:
@guozhangwang thanks for the review.  I'll move the fix to the `StreamPartitionAssignor#onAssignment` .  

From what I can see we'll need to leave the to call to `TopologyBuilder.updateSubscription` in `StreamPartitionAssignor#subscription()` as it updates the leader's topology  used in the`StreamPartitionAssignor.assign` method.

guozhangwang commented:
> From what I can see we'll need to leave the to call to TopologyBuilder.updateSubscription in StreamPartitionAssignor#subscription() as it updates the leader's topology used in theStreamPartitionAssignor.assign method.

makes sense.

bbejeck committed:
KAFKA-5226: Updates per comments

bbejeck commented:
@guozhangwang updates per comments.
   

- Put `SubscriptionUpdates` back as inner class of `StreamPartitionAssignor`
- Removed integration test and added unit test
- Put the check for updated topics in the `StreamPartitionAssignor#onAssignment`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:
```diff
@@ -606,6 +606,25 @@ public void onAssignment(Assignment assignment) {
+
+
+    private void updateSubscribedTopics(Assignment assignment) {
+        if (streamThread != null && streamThread.builder.sourceTopicPattern() != null) {
```
How could `streamThread` be `null` -- `null` would indicate a bug and we might want to fail hard on that instead of masking it with this check?

bbejeck replied:
you're right it can't, I put that in for unit tests, I'll remove and update the tests

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:
```diff
@@ -606,6 +606,25 @@ public void onAssignment(Assignment assignment) {
+    }
+
+
+    private void updateSubscribedTopics(Assignment assignment) {
```
nit: add `final`
also in L616

bbejeck replied:
Ack

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:
```diff
@@ -606,6 +606,25 @@ public void onAssignment(Assignment assignment) {
+            for (TopicPartition topicPartition : assignment.partitions()) {
+                assignedTopics.add(topicPartition.topic());
+            }
+            if (!assignedTopics.isEmpty() && !streamThread.builder.subscriptionUpdates().getUpdates().containsAll(assignedTopics)) {
```
Would `containsAll` not return `true` if `assignedTopics` is empty? (ie, can we remove the first check?)

bbejeck replied:
Good catch, `containsAll` returns if `assignedTopics` is empty, removing first check.

mjsax reviewed (commented):
Just some nits. Overall LGTM.

bbejeck committed:
KAFKA-5226: Updates per comments

bbejeck commented:
@mjsax removed null check for `streamThread` and first check for empty `assignedTopic`

cc\ @guozhangwang

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:
```diff
@@ -606,6 +606,25 @@ public void onAssignment(Assignment assignment) {
+    }
+
+
+    private void updateSubscribedTopics(final Assignment assignment) {
```
nit: could we change this to `maybeUpdateSubscribedTopics(Set<String> topics)` with the check on `streamThread.builder.sourceTopicPattern() != null` within the function so that both `subscription()` and  `onAssignment` could consolidate on this function?

bbejeck replied:
Yes.  This will cause a rebuild of the topology in those cases where all topics are already known. But thinking about it some more, IMHO it's worth the tradeoff for simplifying the logic.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignorTest.java:
```diff
@@ -104,6 +106,12 @@
     private final MockClientSupplier mockClientSupplier = new MockClientSupplier();
     private final TopologyBuilder builder = new TopologyBuilder();
     private final StreamsConfig config = new StreamsConfig(configProps());
+    private final StreamThread mockStreamThread = new StreamThread(builder, config,
```
Why modify this test class to let `partitionAssignor` set this mock stream thread?

bbejeck replied:
There are some tests using `onAssign` which were getting NPEs.  Since in practice the embedded `streamThread` should never be null, I opted to update the test.

guozhangwang replied:
Thanks.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1499,6 +1503,93 @@ protected StreamTask createStreamTask(final TaskId id, final Collection<TopicPar
+        assignment = new PartitionAssignor.Assignment(topicPartitions, info.encode());
+        partitionAssignor.onAssignment(assignment);
+
+        assertTrue(nodeToSourceTopics.get("source").size() == 3);
```
Is the third step (i.e. from 2 topics to 3 topics) intended for additional test coverage?

bbejeck replied:
Yes I wanted to exercise the functionality more than once.

guozhangwang reviewed (commented):
Thanks for the update! LGTM overall, just a few minor comments.

bbejeck committed:
KAFKA-5226: Updates per comments

bbejeck commented:
@guozhangwang thanks for the additional pass, updates per comments.

bbejeck commented:
retest this please

bbejeck commented:
previous failure was unrelated to this PR

bbejeck commented:
previous failure was unrelated to this PR

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1499,6 +1503,93 @@ protected StreamTask createStreamTask(final TaskId id, final Collection<TopicPar
+        final Field
+            nodeToSourceTopicsField =
+            topologyBuilder.getClass().getDeclaredField("nodeToSourceTopics");
+        nodeToSourceTopicsField.setAccessible(true);
```
Arg. Reflections. Seems like our code is not well structure if we need this. Can you file a Jira to do some follow up cleanup (or if this does not make it into 0.11, we might to the cleanup first).

bbejeck replied:
Ack

mjsax reviewed (approved):
LGTM. (beside the reflection comment, but this should not block the PR).

bbejeck committed:
KAFKA-5226: Only update topics (hence rebuild topology) if new topics discovered.

bbejeck commented:
@guozhangwang @mjsax updated per comments and discussion on slack. 

Specifically, now we'll only rebuild the topology under two conditions

1. The topics passed into `StreamPartitionAssignor.subscription` don't exactly match what the current `TopologyBuilder.subscriptionUpdates` contains
2. Topics passed into `StreamPartitionAssignor.onAssign` have newly discovered topics from `ConsumerCoordinator.onJoinComplete`

guozhangwang commented:
LGTM! Merged to trunk and 0.11.0.
